File: sdk/npm/src/providers/etempmail.ts
import { InternalEmailInfo, Email, Channel } from '../types';
import { normalizeEmail } from '../normalize';
import { fetchWithTimeout } from '../retry';

const CHANNEL: Channel = 'etempmail';
const BASE = 'https://etempmail.com';

const BROWSER_HEADERS: Record<string, string> = {
  Accept: '*/*',
  'Accept-Language': 'zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6',
  'Cache-Control': 'no-cache',
  DNT: '1',
  Origin: BASE,
  Pragma: 'no-cache',
  Referer: `${BASE}/zh`,
  'sec-ch-ua': '"Chromium";v="146", "Not-A.Brand";v="24", "Microsoft Edge";v="146"',
  'sec-ch-ua-mobile': '?0',
  'sec-ch-ua-platform': '"Windows"',
  'sec-fetch-dest': 'empty',
  'sec-fetch-mode': 'cors',
  'sec-fetch-site': 'same-origin',
  'User-Agent':
    'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0',
  'X-Requested-With': 'XMLHttpRequest',
};

/** 从响应头提取 `name=value` Cookie 串（支持多条 Set-Cookie） */
function cookieHeaderFromResponse(headers: Headers): string {
  const h = headers as Headers & { getSetCookie?: () => string[] };
  const lines = typeof h.getSetCookie === 'function' ? h.getSetCookie() : [];
  const parts: string[] = [];
  for (const line of lines) {
    const nv = line.split(';')[0].trim();
    if (nv) parts.push(nv);
  }
  if (!parts.length) {
    const single = headers.get('set-cookie');
    if (single) {
      for (const segment of single.split(/,(?=[^ =]+=)/)) {
        const nv = segment.split(';')[0].trim();
        if (nv) parts.push(nv);
      }
    }
  }
  return parts.join('; ');
}

function mergeCookieHeader(existing: string, headers: Headers): string {
  const incoming = cookieHeaderFromResponse(headers);
  if (!incoming) return existing;
  const map = new Map<string, string>();
  for (const chunk of [existing, incoming]) {
    for (const pair of chunk.split(';')) {
      const trimmed = pair.trim();
      if (!trimmed) continue;
      const eq = trimmed.indexOf('=');
      if (eq <= 0) continue;
      const name = trimmed.slice(0, eq).trim();
      const value = trimmed.slice(eq + 1).trim();
      map.set(name, value);
    }
  }
  return [...map.entries()].map(([k, v]) => `${k}=${v}`).join('; ');
}

async function establishSession(): Promise<string> {
  const res = await fetchWithTimeout(`${BASE}/zh`, {
    method: 'GET',
    headers: BROWSER_HEADERS,
  });
  if (!res.ok) {
    throw new Error(`etempmail: session page failed: ${res.status}`);
  }
  const cookie = cookieHeaderFromResponse(res.headers);
  if (!cookie.includes('ci_session=')) {
    throw new Error('etempmail: no ci_session cookie from server');
  }
  return cookie;
}

interface GetEmailAddressResponse {
  id?: string;
  address?: string;
  creation_time?: string;
  recover_key?: string;
}

interface InboxItem {
  subject?: string;
  from?: string;
  date?: string;
  body?: string;
}

export async function generateEmail(): Promise<InternalEmailInfo> {
  let cookie = await establishSession();
  const res = await fetchWithTimeout(`${BASE}/getEmailAddress`, {
    method: 'POST',
    headers: {
      ...BROWSER_HEADERS,
      Cookie: cookie,
    },
  });
  cookie = mergeCookieHeader(cookie, res.headers);
  const rawText = await res.text();
  if (!res.ok) {
    throw new Error(`etempmail: getEmailAddress failed: ${res.status}`);
  }
  let data: GetEmailAddressResponse;
  try {
    data = JSON.parse(rawText) as GetEmailAddressResponse;
  } catch {
    throw new Error(`etempmail: getEmailAddress returned non-JSON: ${rawText.slice(0, 120)}`);
  }
  if (!data.address) {
    throw new Error('etempmail: no address in response');
  }
  const creationSec = data.creation_time ? parseInt(data.creation_time, 10) : NaN;
  return {
    channel: CHANNEL,
    email: data.address,
    token: cookie,
    createdAt: !Number.isNaN(creationSec) ? new Date(creationSec * 1000).toISOString() : undefined,
  };
}

export async function getEmails(recipientEmail: string, cookie: string): Promise<Email[]> {
  const res = await fetchWithTimeout(`${BASE}/getInbox`, {
    method: 'POST',
    headers: {
      ...BROWSER_HEADERS,
      Cookie: cookie,
    },
  });
  const rawText = await res.text();
  if (!res.ok) {
    throw new Error(`etempmail: getInbox failed: ${res.status}`);
  }
  let items: InboxItem[];
  try {
    items = JSON.parse(rawText) as InboxItem[];
  } catch {
    throw new Error(`etempmail: getInbox returned non-JSON: ${rawText.slice(0, 120)}`);
  }
  if (!Array.isArray(items)) {
    throw new Error('etempmail: getInbox expected JSON array');
  }
  return items.map((raw, index) =>
    normalizeEmail(
      {
        id: [raw.from, raw.subject, raw.date, String(index), recipientEmail].join('\n'),
        from: raw.from ?? '',
        subject: raw.subject ?? '',
        body: raw.body ?? '',
        date: raw.date,
        isRead: false,
      },
      recipientEmail,
    ),
  );
}

File: sdk/python/tempmail_sdk/providers/etempmail.py
"""
eTempMail — https://etempmail.com
先 GET /zh 建立 ci_session，再 POST /getEmailAddress、/getInbox。
"""

from typing import List

from .. import http as tm_http
from ..normalize import normalize_email
from ..types import Email, EmailInfo

CHANNEL = "etempmail"
BASE = "https://etempmail.com"

HEADERS = {
    "Accept": "*/*",
    "Accept-Language": "zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6",
    "Cache-Control": "no-cache",
    "DNT": "1",
    "Origin": BASE,
    "Pragma": "no-cache",
    "Referer": f"{BASE}/zh",
    "sec-ch-ua": '"Chromium";v="146", "Not-A.Brand";v="24", "Microsoft Edge";v="146"',
    "sec-ch-ua-mobile": "?0",
    "sec-ch-ua-platform": '"Windows"',
    "sec-fetch-dest": "empty",
    "sec-fetch-mode": "cors",
    "sec-fetch-site": "same-origin",
    "User-Agent": (
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
        "(KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0"
    ),
    "X-Requested-With": "XMLHttpRequest",
}


def _valid_generated_address(address: str) -> bool:
    if not address or len(address) > 254 or address.strip() != address:
        return False
    if address.count("@") != 1:
        return False
    local, domain = address.split("@", 1)
    if not local or len(local) > 64:
        return False
    if local.startswith(".") or local.endswith(".") or ".." in local:
        return False
    if any(ch.isspace() or ord(ch) < 32 or ord(ch) == 127 for ch in local):
        return False
    return _valid_domain(domain)


def _valid_domain(domain: str) -> bool:
    if not domain or len(domain) > 253 or domain.startswith(".") or domain.endswith("."):
        return False
    labels = domain.split(".")
    if len(labels) < 2:
        return False
    for label in labels:
        if not label or len(label) > 63:
            return False
        if label.startswith("-") or label.endswith("-"):
            return False
        if not label.isascii():
            return False
        if any(not (ch.isalnum() or ch == "-") for ch in label):
            return False
    return True


def _cookie_header_from_response(r) -> str:
    parts = []
    for c in r.cookies:
        parts.append(f"{c.name}={c.value}")
    return "; ".join(parts)


def _merge_cookie_header(existing: str, r) -> str:
    m = {}
    for seg in existing.split(";"):
        seg = seg.strip()
        if not seg or "=" not in seg:
            continue
        k, _, v = seg.partition("=")
        k, v = k.strip(), v.strip()
        if k:
            m[k] = v
    for c in r.cookies:
        m[c.name] = c.value
    return "; ".join(f"{k}={v}" for k, v in m.items())


def generate_email() -> EmailInfo:
    r = tm_http.get(f"{BASE}/zh", headers=HEADERS, timeout=15)
    r.raise_for_status()
    cookie = _cookie_header_from_response(r)
    if "ci_session=" not in cookie:
        raise RuntimeError("etempmail: no ci_session cookie")
    h2 = dict(HEADERS)
    h2["Cookie"] = cookie
    r2 = tm_http.post(f"{BASE}/getEmailAddress", headers=h2, data=b"", timeout=15)
    r2.raise_for_status()
    cookie = _merge_cookie_header(cookie, r2)
    data = r2.json()
    if not isinstance(data, dict):
        raise RuntimeError("etempmail: invalid create response")
    addr = data.get("address")
    if not addr:
        raise RuntimeError("etempmail: no address in response")
    if not isinstance(addr, str) or not _valid_generated_address(addr):
        raise RuntimeError("etempmail: invalid address in response")
    created = None
    ct = data.get("creation_time")
    if ct is not None:
        try:
            sec = int(str(ct))
            if sec > 0:
                from datetime import datetime, timezone
                created = datetime.fromtimestamp(sec, tz=timezone.utc).isoformat()
        except (ValueError, OSError):
            pass
    return EmailInfo(channel=CHANNEL, email=str(addr), _token=cookie, created_at=created)


def get_emails(email: str, cookie: str) -> List[Email]:
    if not cookie:
        raise ValueError("token is required for etempmail channel")
    h = dict(HEADERS)
    h["Cookie"] = cookie
    r = tm_http.post(f"{BASE}/getInbox", headers=h, data=b"", timeout=15)
    r.raise_for_status()
    rows = r.json()
    if not isinstance(rows, list):
        return []
    out: List[Email] = []
    for i, raw in enumerate(rows):
        if not isinstance(raw, dict):
            continue
        from_addr = raw.get("from") or ""
        subj = raw.get("subject") or ""
        dt = raw.get("date") or ""
        body = raw.get("body") or ""
        flat = {
            "id": f"{from_addr}\n{subj}\n{dt}\n{i}\n{email}",
            "from": from_addr,
            "subject": subj,
            "body": body,
            "date": dt,
            "isRead": False,
        }
        out.append(normalize_email(flat, email))
    return out
